client/asset/eth/eth.go

```diff
@@ -2402,18 +2526,474 @@ func (eth *ETHWallet) checkForNewBlocks(reportErr func(error)) {
+	contractVer, secretHash, err := dexeth.DecodeContractData(redemption.Spends.Contract)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode contract data: %w", err)
+	}
+	hdr, err := w.node.bestHeader(w.ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get best header: %w", err)
+	}
+	currentTip := hdr.Number.Uint64()
+
+	tx, txBlock, err := w.node.getTransaction(w.ctx, txHash)
+	if errors.Is(err, asset.CoinNotFoundError) {
+		w.log.Errorf("ConfirmRedemption: geth could not find tx: %s", txHash)
+		swapIsRedeemed, err := w.swapIsRedeemed(secretHash, contractVer)
+		if err != nil {
+			return nil, err
+		}
+		if swapIsRedeemed {
+			return confStatus(txConfsNeededToConfirm, txHash), nil
+		}
+
+		// If geth cannot find the transaction, and it also wasn't among the
+		// monitored txs, we will resubmit the swap individually.
+		txs, _, _, err := w.Redeem(&asset.RedeemForm{
+			Redemptions:   []*asset.Redemption{redemption},
+			FeeSuggestion: w.FeeRate(),
+		}, nil, nil)
+		if err != nil {
+			return nil, err
+		}
+		if len(txs) == 0 {
+			return nil, errors.New("no txs returned while resubmitting redemption")
+		}
+		var resubmittedTxHash common.Hash
+		copy(resubmittedTxHash[:], txs[0])
+		return confStatus(0, resubmittedTxHash), nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	var confirmations uint64
+	if txBlock > 0 {
+		if currentTip >= uint64(txBlock) {
+			confirmations = currentTip - uint64(txBlock) + 1
+		}
+	}
+	if confirmations >= txConfsNeededToConfirm {
+		receipt, err := w.node.transactionReceipt(w.ctx, txHash)
+		if err != nil {
+			return nil, err
+		}
+		if receipt.Status == types.ReceiptStatusSuccessful {
+			return confStatus(txConfsNeededToConfirm, txHash), nil
+		}
+		replacementTxHash, err := w.resubmitRedemption(tx, contractVer, nil, feeWallet, nil)
+		if err != nil {
+			return nil, err
+		}
+		return confStatus(0, *replacementTxHash), nil
+	}
+	if confirmations > 0 {
+		return confStatus(confirmations, txHash), nil
+	}
+
+	return w.checkUnconfirmedRedemption(secretHash, contractVer, txHash, tx, currentTip, feeWallet, nil)
+}
+
+// confirmRedemption checks the confirmation status of a redemption transaction.
+// It will resubmit transactions if it has been determined that the transaction
+// cannot be mined.
+func (w *assetWallet) confirmRedemption(coinID dex.Bytes, redemption *asset.Redemption, feeWallet *assetWallet) (*asset.ConfirmRedemptionStatus, error) {
+	if len(coinID) != common.HashLength {
+		return nil, fmt.Errorf("expected coin ID to be a transaction hash, but it has a length of %d",
+			len(coinID))
+	}
+	var txHash common.Hash
+	copy(txHash[:], coinID)
+
+	monitoredTx, err := w.getLatestMonitoredTx(txHash)
+	if err != nil {
+		w.log.Error(err)
+		return w.confirmRedemptionWithoutMonitoredTx(txHash, redemption, feeWallet)
+	}
+	// This mutex is locked inside of getLatestMonitoredTx.
+	defer monitoredTx.mtx.Unlock()
```

Comment:
Consider the following scenario.
When `getLatestMonitoredTx` returned, it unlocked the `monitoredTxnsMtx`, but left `monitoredTx.mtx` locked.
Now assume there's another goroutine calling `confirmRedemption`>`getLatestMonitoredTx` again for this tx for some reason (it's entirely up to the caller). It will get the `monitoredTxnsMtx` successfully and block on `monitoredTx.mtx` while this `confirmRedemption` does its work.   But that can get to `resubmitRedemption`>`recordReplacementTx`, which will attempt `w.monitoredTxsMtx.Lock()`, and it will be deadlock because the two goroutines are waiting on the other's lock.  This is what I was getting at regarding lock ordering leading to deadlock.

Reply:
Yes you're right. I've updated to get the lock and release after each individual access in `getLatestMonitoredTx`.